File: 1235.maximum-profit-in-job-scheduling.cpp
/*
 * @lc app=leetcode id=1235 lang=cpp
 *
 * [1235] Maximum Profit in Job Scheduling
 */
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
#define START 0
#define END 1
#define PROFIT 2

typedef tuple<int, int, int> job;

class Solution {
public:
    // T: O(n * log(n))
    // S: O(n)
    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {
        const int n = startTime.size();
        vector<job> jobs(n);
        for (int i = 0; i < n; ++i) {
            jobs[i] = {startTime[i], endTime[i], profit[i]};
        }

        sort(jobs.begin(), jobs.end(), [](const job &a, const job &b) { 
            return get<END>(a) < get<END>(b);   // Sort by end time ascendingly
        });
        
        vector<int> dp(n, 0);   // dp[i] represents maximum profit for jobs[:i]
        dp[0] = get<PROFIT>(jobs[0]);   // Base case: the profit of the job ending first
        for (int i = 1; i < n; ++i) {
            // Base case: select just the current job or the previous selected job(s)
            dp[i] = max(dp[i - 1], get<PROFIT>(jobs[i]));

            // Try including more jobs
            int j = i - 1;
            while (j >= 0 && get<END>(jobs[j]) > get<START>(jobs[i]))  --j;
            // Since jobs are sorted by end time, find the closest previous job 
            // without overlapping the current job, and decide whether to 
            // add profits from these ealier jobs or not
            if (j >= 0)  dp[i] = max(dp[i], get<PROFIT>(jobs[i]) + dp[j]);
        }
        return dp.back();
    }
};
// @lc code=end

File: 1235.maximum-profit-in-job-scheduling.py
#
# @lc app=leetcode id=1235 lang=python3
#
# [1235] Maximum Profit in Job Scheduling
#
from typing import *

# @lc code=start
class Solution:
    # T: O(n ^ 2)
    # S: O(n)
    def jobScheduling(self, startTime: List[int], endTime: List[int], profit: List[int]) -> int:
        n = len(startTime)

        # Sort by start and then end
        jobs = sorted([(startTime[i], endTime[i], profit[i]) for i in range(n)])
        
        dp = dict()

        def memo_dfs(i):
            if i == n:
                return 0
            if i in dp:
                return dp[i]

            # Option 1: take the current job, and find the next available one
            res_incl = jobs[i][2]
            next_i = self.find_next_job(jobs, i)
            if next_i != -1:
                res_incl += memo_dfs(next_i)

            # Option 2: skip the current job
            res_excl = memo_dfs(i + 1)

            res = max(res_incl, res_excl)
            dp[i] = res
            return res

        return memo_dfs(0)

    @staticmethod
    def find_next_job(jobs: List[Tuple], i: int) -> int:
        for j in range(i + 1, len(jobs)):
            if jobs[j][0] >= jobs[i][1]:
                # Next start time greater than current end time,
                # no overlap.
                return j
        return -1

# @lc code=end

